### PR TITLE
[oneDPL] + add comments for own implementation instead of std::gcd and std::lcm inside oneDPL code

### PR DESCRIPTION
**Pull request by SergeyKopienko**
We made investigation why we need have own implementation instead of std::gcd and std::lcm inside oneDPL code.

### CODE REVIEW

File: include/oneapi/dpl/numeric [hunk 1]
```diff
@@ -39,6 +38,7 @@ namespace oneapi
 namespace dpl
 {
 using ::std::accumulate;
+using ::std::gcd;
```
Comment (akukanov) [inline include/oneapi/dpl/numeric:41]: what about `lcm`? :)

**Comment by akukanov** [comment]
Let's start writing some meaningful descriptions to PRs. What's the motivation for this change, and why we think it's appropriate? Ruslan asked: "Is there any precedence for those functions not being in modern C++ compilers with C++17 support?" and I see no answer to this question here.

File: include/oneapi/dpl/numeric (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/numeric:41]: Added

File: include/oneapi/dpl/internal/gcd_impl.h [hunk 2]
```diff
@@ -55,6 +55,11 @@ __get_abs(_Source __t)
 } // namespace internal
 
 // gcd
+// We need have own implementation instead of std::gcd because
```
Comment (rarutyun) [inline include/oneapi/dpl/internal/gcd_impl.h:58]: Is not it better to make one comment at the beginning of the header for both `gcd` and `lcm`?

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/gcd_impl.h]: Fixed

File: include/oneapi/dpl/internal/gcd_impl.h [hunk 3]
```diff
@@ -27,6 +27,12 @@ namespace dpl
 
 namespace internal
 {
+// We need have own implementation instead of std::gcd and std::lcm because
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/gcd_impl.h:30]: 1) I would place the comment directly before gcd and lcm implementations.
2) I think we should write about an error.. A short explanation why and "reference" (not link) to the C++ limitations in SYCL spec.
f.e. "SYCL doesn't  support recursive call in a kernel. For more details about C++ limitations see the SYCL specification."

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/gcd_impl.h]: Fixed

Comment (akukanov) [reply include/oneapi/dpl/internal/gcd_impl.h]: I think Mikhail meant to keep a single comment, but put it right before the function implementation starts. At least, I see no reason to duplicate the exact same wording twice. I would keep only the first instance, and moreover put a space line after the comment to visually separate it from the following function.

As for the wording, I would slightly change it, e.g. "Some C++ standard libraries implement std::gcd and std::lcm as recursive functions, which prevents their use in SYCL kernels (see the SYCL specification for more details). Therefore oneDPL provides its own implementation."

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/gcd_impl.h]: Fixed

**Comment by SergeyKopienko** [comment]
@rarutyun, @MikeDvorskiy, @akukanov could you please approve this PR ?

**Comment by rarutyun** [comment]
> @rarutyun, @MikeDvorskiy, @akukanov could you please approve this PR ?

Sorry, I cannot, because from my understanding that's not what we agreed on. During the discussion we said that there are two ways of writing those comments:
 - in-place (for every function)
 - one comment for the header.
I believe Sergey, you said that you want to write comments in place, but what I see in the something intermediate between to approaches I've described above.

You have one comment but it's at the beginning of the header it's somewhere. When I proposed the wording for the comments I assumed it's supposed to be in-place. But in that case I would prefer having two comments (one per function).

If you put a comment as it is at the beginning of the header (effectively making it one per header) I think the wording should be slightly modified.

**Comment by SergeyKopienko** [comment]
> > @rarutyun, @MikeDvorskiy, @akukanov could you please approve this PR ?
> 
> Sorry, I cannot, because from my understanding that's not what we agreed on. During the discussion we said that there are two ways of writing those comments:
> 
> * in-place (for every function)
> * one comment for the header.
>   I believe Sergey, you said that you want to write comments in place, but what I see in the something intermediate between to approaches I've described above.
> 
> You have one comment but it's at the beginning of the header it's somewhere. When I proposed the wording for the comments I assumed it's supposed to be in-place. But in that case I would prefer having two comments (one per function).
> 
> If you put a comment as it is at the beginning of the header (effectively making it one per header) I think the wording should be slightly modified.

Current comment state was made by me like @akukanov proposed, as far as I understood his proposal.

**Comment by akukanov** [comment]
> Current comment state was made by me like @akukanov proposed, as far as I understood his proposal.

It was Mikhail's proposal, actually, as he said "I would place the comment directly before gcd and lcm implementations.". But yes, the patch matches my understanding of that. And I think that the exact placement of the comment(s) is not really so important. and is not worth further discussion. So I am OK with this patch and approve.

**Review by akukanov — APPROVED** [review]

**Pull request merged**